beefy-gadget/src/gossip.rs

```diff
@@ -116,33 +157,35 @@ where
 	}
 
 	fn message_expired<'a>(&'a self) -> Box<dyn FnMut(B::Hash, &[u8]) -> bool + 'a> {
-		let live_rounds = self.live_rounds.read();
+		let known_votes = self.known_votes.read();
 		Box::new(move |_topic, mut data| {
 			let msg = match VoteMessage::<MmrRootHash, NumberFor<B>, Public, Signature>::decode(&mut data) {
 				Ok(vote) => vote,
 				Err(_) => return true,
 			};
 
-			let expired = !GossipValidator::<B>::is_live(&live_rounds, msg.commitment.block_number);
+			let round = msg.commitment.block_number;
+			let expired = !GossipValidator::<B>::is_live(&known_votes, &round);
 
-			trace!(target: "beefy", "🥩 Message for round #{} expired: {}", msg.commitment.block_number, expired);
+			trace!(target: "beefy", "🥩 Message for round #{} expired: {}", round, expired);
 
 			expired
 		})
 	}
 
 	#[allow(clippy::type_complexity)]
 	fn message_allowed<'a>(&'a self) -> Box<dyn FnMut(&PeerId, MessageIntent, &B::Hash, &[u8]) -> bool + 'a> {
-		let live_rounds = self.live_rounds.read();
+		let known_votes = self.known_votes.read();
 		Box::new(move |_who, _intent, _topic, mut data| {
```

Comment:
According to how [propage()](https://github.com/paritytech/substrate/blob/cf4e320398cf3b8ef8a2c240d7495c25b3b73b08/client/network-gossip/src/state_machine.rs#L131-L138) handles `MessageIntent::PeriodicRebroadcast`, I'm inclined to just drop rebroadcasts. Something like

```rust
		Box::new(move |_who, intent, _topic, mut data| {
			// We drop rebroadcasts. If a peer doesn't know a message, the gossip state machine will
			// treat a rebroadcast as a broadcast. In other words, if the do see a rebroadcast, it is
			// *alwayas* an already known message.
			if intent == MessageIntent::PeriodicRebroadcast {
				return false;
			}

```
I tested this with `rococo-local` and it worked well. Theoretically this should give us a pretty minimal amount of gossip messages.

Reply:
Hmm, but this doesn't guarantee that the round is concluded for everyone really. Until we have nodes requesting missing BEEFY justifications (I guess part of import queue stuff) I think we should be on the safe side of gossiping more, but without risking some sessions not being concluded.

I guess it is a potential workaround for now as well, but given this cache receiving re-broadcasted votes should be cheap.

Reply:
Totally agree, regarding this being an interim solution. Right now, our problem is, we Gossip *way* to much. I think we should give it a try and see how it goes. Right now, we don't produce any BEEFY justifications on Rococo. Guess, it won't be worse than that.